docs/dev/raft-topology.md

```diff
@@ -0,0 +1,115 @@
+        throw an error
+
+def topology_change():
+    isolate_nodes(ignored_nodes)
+    version += 1
+    ring_slice.state = write_both_read_old
+    # The purpose of the barrier is to make sure that there are no writes to old replicas only.
+    # All new writes are to old+new or fenced after this.
+    # Only write requests should be fenced (though it is fine to fence both).
+    # No need to drain writes on replicas since only writes that completed on a coordinator have to be
+    # captured by the streaming
+    affected_nodes = old replica set
+    raft.write(rind_slice.state + version);
+    try:
+        global_token_metadata_barrier(affected_nodes)
+        # Now at least the old replica set knows the new topology and fences version-1.
+        # No uncompleted unfenced writes to only old replica set are in flight
+        stream()
+    except:
+        # If we are hear we did not manage to stream yet and thus should go back to old topology
+        # Some or all nodes may write to both replica sets new and old, reads are from the old one
+        version += 1
+        revert_token_metadata_changes()
+        raft.write(token metadata state + version)
+        cleanup(new replica set)
+        return
+
+    version += 1
+    ring_slice.state = write_both_read_new
+    # The purpose of the barrier is to make sure no unfenced reads are going to the old
+    # owner. Only reads should be fenced (though it is fine to fence both).
+    # No need to drain writes on replicas.
+    affected_nodes = old replica set
+    raft.write_state(state + version)
+    Try:
+        # ignore-dead-nodes concept should be a system table managed by group0. We should accept updates
+        # to it either via CQL or via nodetool api: it should be possible to extend the dead node list.
+        # If the barrier fails, we report about the failure into the log and wait until ignore-dead-nodes
+        # list is extended by the DBA, by periodically re-reading the contents of the ignore-dead-nodes
+        # table. As soon as we get a new node we can isolate, we isolate it and retry. We refuse to
+        # isolate a node if it will lead to a loss of majority. In this case the dba should restar
+        # or otherwise enable the dead node.
+        global_token_metadata_barrier(affected_nodes)
+    except:
+        # if we are here it means one or more of the nodes that are losing range cannot be accessed
+        # (at least from the coordinator)
+        # Some nodes are reading from new replica set, but some may still read from the old one
+        # If nodes are really dead we can just proceed since there is no need to enable fence on
+        # a dead node, but may be those nodes are accessible by other nodes in the cluster that may
+        # still read from them let's try to isolate them and continue
+        Try:
+            isolate(dead nodes from the set of nodes that are losing range) # what to do if this also fails?
```

Comment:
We cannot decide to isolate nodes because we don't know if they're permanently dead. Only admin can do this.

Reply:
'Isolation' is not permanent. The node can rejoin the cluster after it syncs the topology. The doc omits how exactly isolation should be implemented. It cannot be done for all verbs on messaging level since we want raft to work to be able to catch up with the topology.

Reply:
> Doesn't it mean that a temporary failure to communicate with them will turn it into a permanent failure which requires admin intervention to fix? This can impact cluster availability.

Not necessary. Are node may see that it was isolated (this will be a flag in raft) and do recovery automatically. But this series has nothing to do with it. This should be discussed on the spec the test from copied from.

Reply:
Maybe the PR shouldn't document algorithm which isn't implemented yet.